### PR TITLE
feat: add awesome-copilot plugin format adapters, schema, and validation

### LINKED ISSUE #277: [Feature]: Plugins support for VSCode
### Problem Statement

As a user of prompt registry I want to be able to install plugins with it

### Proposed Solution

TDB

### Alternatives Considered

_No response_

### Examples

_No response_

### Component

None

### Priority (from your perspective)

Low - Nice to have

### Contribution

- [ ] I would be willing to submit a PR for this feature
- [ ] I can help with testing
- [ ] I can help with documentation

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by wherka-ama**
## Description

Support the new plugin directory structure from awesome-copilot PR #717. Add remote and local adapters for plugins/<id>/.github/plugin/plugin.json, JSON schema for plugin manifest validation, and a workspace validation command.
 
## Type of Change

- [x] ✨ New feature (non-breaking change which adds functionality)
- [x] 📝 Documentation update

## Related Issues

Relates to [#200](https://github.com/AmadeusITGroup/prompt-registry/issues/200)

## Changes Made

- Add AwesomeCopilotPluginAdapter (remote) and LocalAwesomeCopilotPluginAdapter
- Add plugin.schema.json with support for agents, skills, MCP servers, display
- Add validatePlugin() to SchemaValidator
- Add ValidatePluginsCommand for workspace plugin validation
- Register new source types and adapters in RegistryManager
- Add source UI options in SourceCommands

### Test Coverage

- [x] Unit tests added/updated
- [x] Integration tests added/updated
- [x] Manual testing completed
- [x] All existing tests pass

### Manual Testing Steps

1. Add new source of the type `Awesome Copilot Plugins`
2. Point at https://github.com/github/awesome-copilot , main(branch), plugins(folder)
3. Synchronise the new source
4. Verify the Marketplace - it should display the tiles related to the content from the new source
5. Install selected plugin
6. Verify the details after installation
7. Verify that the primitives are available in the context

### Tested On

- [ ] Linux

- [x] VS Code Stable
- [x] VS Code Insiders

## Screenshots

image

image

image

image

image

image

## Checklist

- [x] My code follows the project's style guidelines
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings or errors
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published

## Documentation

- [x] README.md updated
- [x] JSDoc comments added/updated

---

**By submitting this pull request, I confirm that my contribution is made under the terms of the Apache License 2.0.**

### CODE REVIEW

File: docs/author-guide/plugin-schema.md [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+}
+```
+
+## Item Kinds
+
+| `kind` | Mapped to `type` in deployment manifest | Typical path |
+|--------|-----------------------------------------|--------------|
+| `prompt` | `prompt` | `./prompts/*.prompt.md` |
+| `instruction` | `instructions` | `./instructions/*.instructions.md` |
+| `chat-mode` | `chatmode` | `./chatmodes/*.chatmode.md` |
```
Comment (gblanc-1a) [inline docs/author-guide/plugin-schema.md]: chat-mode are obsolete and do not exit anymore

File: docs/author-guide/adding-profile-source-to-hub.md [hunk 2]
```diff
@@ -41,6 +41,7 @@ Choose the appropriate source type for organization we recommend github source t
 |------|----------|-----------------|
 | `github` | GitHub repository | `repository` |
 | `awesome-copilot` | YAML collections | `repository` |
+| `awesome-copilot-plugin` | JSON plugins (`plugin.json` format) | `repository` |
```
Comment (gblanc-1a) [inline docs/author-guide/adding-profile-source-to-hub.md:44]: Not sure if we want to put it right away in the author-guide as we wanted to be first in consumption mode

File: src/adapters/awesome-copilot-plugin-adapter.ts [hunk 3]
```diff
@@ -0,0 +1,624 @@
...
+      this.logger.info('[AwesomeCopilotPluginAdapter] Using explicit token from configuration');
+      return this.authToken;
+    }
+
+    this.authMethod = 'none';
+    this.logger.warn('[AwesomeCopilotPluginAdapter] No authentication available');
+    return undefined;
+  }
+
+  private async fetchUrl(url: string, redirectDepth = 0): Promise<string> {
```
Comment (gblanc-1a) [inline src/adapters/awesome-copilot-plugin-adapter.ts]: > **FN1** 🔴 Critical — `src/adapters/awesome-copilot-plugin-adapter.ts:572-620` — Auth token leaked on cross-origin redirects
>
> `fetchUrl()` sends the `Authorization: token ${token}` header on every redirect, including cross-origin ones. When `raw.githubusercontent.com` redirects to a CDN, the GitHub token is forwarded to an arbitrary domain. This is an OWASP-class credential leakage vulnerability (Sensitive Data Exposure).
>
> **Recommendation:** Parse the origin of the redirect URL and strip the `Authorization` header when the redirect leaves `*.github.com` / `*.githubusercontent.com`. Or call `res.resume()` before following the redirect (see FN5).

File: src/commands/validate-plugins-command.ts
Comment (gblanc-1a) [inline src/commands/validate-plugins-command.ts:1]: > **QA2** 🔴 Critical — `src/commands/validate-plugins-command.ts` — No test file for 259-line command
>
> `ValidatePluginsCommand` contains `execute()`, `findPluginDirectories()`, `validatePlugin()`, duplicate-ID detection, `listOnly` mode, schema validation, and item path checks — all without a single test. This is significant untested business logic with multiple branching paths.
>
> **Recommendation:** Create `test/commands/validate-plugins-command.test.ts` covering: no workspace, missing plugins dir, valid plugins, schema errors, duplicate IDs, `listOnly` mode, item path not found, and JSON parse failure.

**Review by gblanc-1a — CHANGES_REQUESTED** [review]
When trying to install a plugin with a MCP it succeed but MCP is not configured
For example: https://github.com/github/awesome-copilot/tree/main/plugins/context-matic

Is there maybe a way to benefit from the plugin installation from vscode itself ?
Would be great if we could see it installed even in the plugin list (but might be not possible to do).
Or maybe we could add the plugin source in the vscode marketplace and then open the view of the vscode marketplace ?

File: schemas/plugin.schema.json [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+  "$id": "https://github.com/AmadeusITGroup/prompt-registry/schemas/plugin.schema.json",
+  "title": "Awesome Copilot Plugin",
+  "description": "Schema for Copilot plugin manifest files (plugin.json). Compatible with github/awesome-copilot plugin format and extended with Prompt Registry features (MCP servers, display preferences).",
+  "type": "object",
+  "required": [
+    "id",
+    "name",
+    "description",
+    "items"
+  ],
```
Comment (gblanc-1a) [inline schemas/plugin.schema.json:9]: **FN2** 🔴 Critical — Schema requires `id`, `description`, and `items` — rejects valid upstream plugins

Per the [official VS Code plugin format](https://code.visualstudio.com/docs/copilot/customization/agent-plugins), only `name` is required. Upstream `github/awesome-copilot` plugins use `agents`/`skills` arrays without an `items` property, often have no `id` field (identity is derived from the directory name), and `description` is optional.

The adapters bridge this via `derivePluginItems()`, but `ValidatePluginsCommand` runs schema validation **first** and will reject all valid upstream-format plugins. Furthermore, `agents` and `skills` are not even defined as properties in this schema.

**Recommendation:**
- Change `"required"` to `["name"]`
- Make `id`, `description`, and `items` optional
- Add `agents` (string or string[]) and `skills` (string or string[]) as first-class optional properties
- Consider a custom validation rule requiring at least one of `items`, `agents`, or `skills`

**Review by gblanc-1a — COMMENTED** [review]
## Code Review — Critical & Major Findings

Detailed review covering security, schema compliance with the [official VS Code plugin format](https://code.visualstudio.com/docs/copilot/customization/agent-plugins), test coverage gaps, and technical debt.

See full report for minor/suggestion-level items.

File: src/adapters/plugin-adapter-shared.ts [hunk 5]
```diff
@@ -0,0 +1,306 @@
+/**
```
Comment (gblanc-1a) [inline src/adapters/plugin-adapter-shared.ts:1]: **QA2** 🔴 Critical — No unit tests for this 306-line pure helper module

11 exported pure functions (`mapKindToType`, `calculateBreakdown`, `derivePluginItems`, `inferEnvironments`, `stripLeadingDotSlash`, `toPosixPath`, `stripMdExtension`, `deriveSimpleItemId`, `titleCase`, `toYaml`, `createDeploymentManifest`) have **zero** dedicated test coverage.

These functions drive the core manifest generation logic. Edge cases (unknown kind, `toYaml` with nested arrays, special characters, mixed-case tags for `inferEnvironments`) are not directly verified.

**Recommendation:** Create `test/adapters/plugin-adapter-shared.test.ts` with targeted unit tests for all exported functions, focusing on edge cases.

File: src/adapters/awesome-copilot-plugin-adapter.ts [hunk 6]
```diff
@@ -0,0 +1,624 @@
...
+      const tags = manifest.tags || manifest.keywords || [];
+
+      const bundle: Bundle = {
+        id: pluginId,
+        name: manifest.name || pluginDir,
+        version: manifest.version || '1.0.0',
+        description: manifest.description,
+        author: manifest.author?.name || this.extractRepoOwner(),
+        repository: manifest.repository || this.source.url,
+        tags,
```
Comment (gblanc-1a) [inline src/adapters/awesome-copilot-plugin-adapter.ts:323]: **FN3** 🟠 Major — `manifest.author?.name` silently loses string author

When `plugin.json` has `"author": "John Doe"` (valid per both the JSON schema's `oneOf: [string, object]` and per the official VS Code format), `.name` on a string returns `undefined`. The author is silently dropped and falls back to the repo owner.

Same issue in `local-awesome-copilot-plugin-adapter.ts:305` (falls back to `'Local Developer'`).

**Recommendation:** Update `PluginManifest.author` type to `string | { name: string; url?: string; email?: string }` and add a helper:
```ts
const authorName = typeof manifest.author === 'string' ? manifest.author : manifest.author?.name;
```

File: src/adapters/plugin-adapter-shared.ts [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+        }
+      } else {
+        result += `${pad}${key}: [${value.map((v) => `"${v}"`).join(', ')}]\n`;
+      }
+    } else if (typeof value === 'object' && value !== null) {
+      result += `${pad}${key}:\n${toYaml(value, indent + 1)}`;
+    } else {
+      result += `${pad}${key}: ${JSON.stringify(value)}\n`;
+    }
+  }
```
Comment (gblanc-1a) [inline src/adapters/plugin-adapter-shared.ts]: **TD1 + FN5** 🟠 Major — Custom `toYaml()` reinvents `js-yaml` (already a direct dependency) and has escaping bugs

**TD1**: The project uses `js-yaml` (`yaml.dump()`) in 20+ files. The inline comment "we avoid pulling in a full YAML library" is incorrect — `js-yaml` is already a direct dependency. The hand-rolled implementation lacks edge-case coverage for special characters (strings containing colons, quotes, newlines, YAML reserved words like `true`/`null`).

**FN5**: The array-of-primitives branch at line 241 uses unescaped template literals `"${v}"`. Tags containing double-quote characters produce malformed YAML: `"my "tag""`. The scalar branch correctly uses `JSON.stringify`, but the array branch does not.

**Recommendation:** Replace `toYaml(manifest)` with `yaml.dump(manifest)` from `js-yaml`. Remove `toYaml` entirely.

File: src/adapters/awesome-copilot-plugin-adapter.ts [hunk 8]
```diff
@@ -0,0 +1,624 @@
...
+                break;
+              }
+            }
+            reject(new Error(errorMsg));
+          }
+        });
+      }).on('error', (error) => reject(error));
+    });
+  }
+}
```
Comment (gblanc-1a) [inline src/adapters/awesome-copilot-plugin-adapter.ts:676]: **TD2** 🟠 Major — ~150 lines of auth/HTTP code duplicated from `AwesomeCopilotAdapter`

`getAuthenticationToken()`, `fetchUrl()`, `buildApiUrl()`, `buildRawUrl()`, `parseGitHubUrl()`, `extractRepoOwner()`, and `forceAuthentication()` are near-identical between the two remote adapters. Future auth fixes (e.g., fine-grained PATs, retry logic, rate-limit headers, the redirect token-leak fix from FN1) need two-site changes.

**Recommendation:** Extract a shared `GitHubApiClient` utility. This could be the start of a bigger refactor later for **all** GitHub interactions in the codebase — the collection adapter, the plugin adapter, and any future GitHub-backed sources would all benefit from a single, well-tested HTTP client with proper auth handling and redirect safety.

File: test/adapters/awesome-copilot-plugin-adapter.test.ts [hunk 9]
```diff
@@ -0,0 +1,696 @@
...
+            entries.set(entry.fileName, Buffer.concat(chunks).toString('utf8'));
+            zipfile.readEntry();
+          });
+          readStream.on('error', reject);
+        });
+      });
+      zipfile.on('end', () => resolve(entries));
+      zipfile.on('error', reject);
+    });
+  });
```
Comment (gblanc-1a) [inline test/adapters/awesome-copilot-plugin-adapter.test.ts]: **QA3** 🟠 Major — `extractZipBuffer` helper duplicated verbatim

This 32-line function is copy-pasted between both test files (`awesome-copilot-plugin-adapter.test.ts:24-55` and `local-awesome-copilot-plugin-adapter.test.ts:25-56`). Per repo convention (`test/AGENTS.md`), test helpers should be shared.

**Recommendation:** Extract into `test/helpers/zip-test-helpers.ts` and import in both test files.

File: src/commands/source-commands.ts [hunk 10]
```diff
@@ -424,7 +496,11 @@ export class SourceCommands {
       // Step 4: Check if private/authentication needed (skip for local sources)
       let token: string | undefined;
       let isPrivate: { label: string; description: string; value: boolean } | undefined;
-      const isLocalSource = sourceType.value === 'local' || sourceType.value === 'local-awesome-copilot' || sourceType.value === 'local-apm';
+      const isLocalSource = sourceType.value === 'local'
+        || sourceType.value === 'local-awesome-copilot'
+        || sourceType.value === 'local-awesome-copilot-plugin'
+        || sourceType.value === 'local-apm'
+        || sourceType.value === 'local-skills';
```
Comment (gblanc-1a) [inline src/commands/source-commands.ts:503]: **QA4** 🟠 Major — No tests for 77 new lines of plugin source type branches

The additions handle `awesome-copilot-plugin` and `local-awesome-copilot-plugin` in `getSourceUrl`, the source type picker, and source info display. Zero test coverage for these new code paths.

**Recommendation:** Add test cases to `test/commands/source-commands.test.ts` covering the new source types.

File: src/adapters/plugin-adapter-shared.ts [hunk 11]
```diff
@@ -0,0 +1,306 @@
+/**
+ * Shared types and pure helpers for the awesome-copilot plugin adapters.
+ *
+ * Both `AwesomeCopilotPluginAdapter` (remote/GitHub) and
+ * `LocalAwesomeCopilotPluginAdapter` (local filesystem) consume the same
+ * `plugin.json` format and produce the same deployment-manifest.yml shape.
+ *
+ * Everything in this module is **pure / I/O-free** — no HTTP, no filesystem —
+ * so it can be unit-tested in isolation and shared between adapters without
+ * coupling them to a specific file source.
```
Comment (gblanc-1a) [inline src/adapters/plugin-adapter-shared.ts:10]: **TD3** 🟡 Minor — Extracted functions still duplicated as private methods in the original adapter

`inferEnvironments`, `calculateBreakdown`, `mapKindToType`, and `titleCase` are correctly extracted here as shared pure functions — great factoring. However, the originals remain as private methods in `AwesomeCopilotAdapter` (`awesome-copilot-adapter.ts` lines 350, 366, 408, 657).

**Recommendation:** Follow-up to migrate the old collection adapter to import from this shared module. This avoids future divergence and reduces the surface area of the original adapter.

File: src/commands/validate-plugins-command.ts [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+    show_badge?: boolean;
+  };
+  external?: boolean;
+  repository?: string;
+  homepage?: string;
+  license?: string;
+}
+
+interface PluginValidationResult {
+  errors: string[];
```
Comment (gblanc-1a) [inline src/commands/validate-plugins-command.ts:16]: **TD5** 🟡 Minor — Re-declares `Plugin`/`PluginItem` interfaces that already exist in `plugin-adapter-shared.ts`

These interfaces duplicate `PluginManifest` and `PluginItem` from `../adapters/plugin-adapter-shared`. This creates two sources of truth — schema changes must be applied in two places.

**Recommendation:** Import `PluginManifest` and `PluginItem` from `../adapters/plugin-adapter-shared` instead of re-declaring them.

File: docs/author-guide/adding-profile-source-to-hub.md (hunk 2 above)
Comment (wherka-ama) [reply docs/author-guide/adding-profile-source-to-hub.md:44]: Ok, we could suppress it here for now.

File: docs/author-guide/plugin-schema.md (hunk 1 above)
Comment (wherka-ama) [reply docs/author-guide/plugin-schema.md]: Indeed. We can clear it.
